include/LightGBM/utils/openmp_wrapper.h

```diff
@@ -5,6 +5,15 @@
 #ifndef LIGHTGBM_OPENMP_WRAPPER_H_
 #define LIGHTGBM_OPENMP_WRAPPER_H_
 
+#include <LightGBM/export.h>
+
+// this can only be changed by LGBM_SetMaxThreads()
+LIGHTGBM_EXTERN_C int LGBM_MAX_NUM_THREADS;
+
+// this is modified by OMP_SET_NUM_THREADS(), for example
+// by passing num_thread through params
+LIGHTGBM_EXTERN_C int LGBM_DEFAULT_NUM_THREADS;
```

Comment:
This makes these process-global variables, and therefore not thread safe. 

Like @david-cortes alluded to in the description of #6152.

For example, if you created 2 `Booster` objects in different threads which had different values of `num_threads` in `Config`, one's `OMP_SET_NUM_THREADS()` call could affect code in the other.

I think that's an acceptable risk for now, in exchange for the other benefits of this PR.

Reply:
maybe you can try `thread_local`, but not hurry in this PR.

Reply:
Thank you! I did see that we have that preprocessor macro set up

https://github.com/microsoft/LightGBM/blob/e797985227a012a837c20eddc457de6b7fc7aeaa/include/LightGBM/utils/log.h#L27-L31

but didn't test it out. Let's save it for a follow-up PR... I think it'd be ok to release this PR's changes without making this configuration thread-safe.